--- a/arch/mips/include/asm/mach-ralink/rt305x.h
+++ b/arch/mips/include/asm/mach-ralink/rt305x.h
@@ -148,6 +148,12 @@ static inline int soc_is_rt5350(void)
 #define RT3352_GPIO_MODE_LNA		18
 #define RT3352_GPIO_MODE_PA		20
 
+#define RT5350_GPIO_MODE_SPI_CS1_MASK	3
+#define RT5350_GPIO_MODE_SPI_CS1_SHIFT	21
+#define RT5350_GPIO_MODE_SPI_CS1	0
+#define RT5350_GPIO_MODE_WDT_RST	1
+#define RT5350_GPIO_MODE_GPIO_27	2
+
 #define RT3352_SYSC_REG_SYSCFG0		0x010
 #define RT3352_SYSC_REG_SYSCFG1         0x014
 #define RT3352_SYSC_REG_CLKCFG1         0x030
--- a/arch/mips/ralink/rt305x.c
+++ b/arch/mips/ralink/rt305x.c
@@ -44,6 +44,10 @@ static struct rt2880_pmx_func rgmii_func
 static struct rt2880_pmx_func rt3352_lna_func[] = { FUNC("lna", 0, 36, 2) };
 static struct rt2880_pmx_func rt3352_pa_func[] = { FUNC("pa", 0, 38, 2) };
 static struct rt2880_pmx_func rt3352_led_func[] = { FUNC("led", 0, 40, 5) };
+static struct rt2880_pmx_func rt5350_spi_cs1_func[] = {
+	FUNC("spi_cs1", RT5350_GPIO_MODE_SPI_CS1, 27, 1),
+	FUNC("wdt_rst", RT5350_GPIO_MODE_WDT_RST, 27, 1),
+};
 
 static struct rt2880_pmx_group rt3050_pinmux_data[] = {
 	GRP("i2c", i2c_func, 1, RT305X_GPIO_MODE_I2C),
@@ -81,6 +85,8 @@ static struct rt2880_pmx_group rt5350_pi
 	GRP("uartlite", uartlite_func, 1, RT305X_GPIO_MODE_UART1),
 	GRP("jtag", jtag_func, 1, RT305X_GPIO_MODE_JTAG),
 	GRP("led", rt5350_led_func, 1, RT5350_GPIO_MODE_PHY_LED),
+	GRP_G("spi_cs1", rt5350_spi_cs1_func, RT5350_GPIO_MODE_SPI_CS1_MASK,
+		RT5350_GPIO_MODE_GPIO_27, RT5350_GPIO_MODE_SPI_CS1_SHIFT),
 	{ 0 }
 };
 
